### PR TITLE
[Merged by Bors] - feat: The Kruskal-Katona theorem

### PR DESCRIPTION
**Pull request by YaelDillies**
Prove the Kruskal-Katona theorem.

From LeanCamCombi

Co-authored-by: Bhavik Mehta <bhavikmehta8@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:237]: This isn't strictly true, since the definition of initial segments of colex that's used below depends on the order of the ground set. What is true, however, is that the *size* of C doesn't depend on the order of the ground set. 
The original motivation for the todo on the order iso between colex and Nat was to make the statement of KK genuinely not depend on the order of the ground set, so I wonder if @apnelson1's recent work makes the statements below nicer.

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:232]: To the Lean 4 repo?

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:237]: I mean, the better way of doing things is the k-cascade representation, right? It's a bit annoying that I lost my work on that due to some git misshap, but I could possibly redo it if you think that making the statement order-independent is so important

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/SetFamily/KruskalKatona.lean]: #15705

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (b-mehta) [reply Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:237]: I think we shouldn't write something in a docstring that's false

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (b-mehta) [reply Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:232]: Is this a Lean 4 issue? That's not clear from the description you give here.

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:166]: weird quotes here?

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:237]: I don't think the docstring is false? It's true that `kruskal_katona` is order-dependent, but it's also true that the (usual) mathematical statement of Kruskal-Katona is not.

Comment (b-mehta) [reply Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:237]: Then let's make sure that's clear in the docstring

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:232]: It's a bit of a mix since it goes through `LinearOrder.liftWithOrd` which is in mathlib, but the original sin is that `instDecidableEqFin` does not agree with `decidable_of_iff`

**Comment by b-mehta** [comment]
Annoying. Let's at least make sure this issue is logged somewhere else -
even if only on zulip - and not just in this file

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:56]: No longer present

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/SetFamily/KruskalKatona.lean]: > Annoying. Let's at least make sure this issue is logged somewhere else - even if only on zulip - and not just in this file

https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/DecidableEq.20diamond.20on.20Fin

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:287]:
```suggestion
  -- Yes. Then apply the smallest compression, then keep going
```

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:276]: tempting to try strong induction on family measure here... but can leave it for now if you want

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:242]: private, probably? familyMeasure is private  and `kruskal_katona_helper` is too

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:325]:
```suggestion
/-- We can strengthen Kruskal-Katona slightly: note the cardinality assumption has been relaxed to a `≤`.
```

**Review by b-mehta — APPROVED** [review]
LGTM. Let's make sure to get another review on this though, again I wrote too much of it!

maintainer merge

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:242]: Rah yes I always assume that lemmas whose type include a private definition are themselves private, but this seems not to be the case.

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/SetFamily/KruskalKatona.lean]: That was easy :)

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/SetFamily/KruskalKatona.lean]: I have opened https://github.com/leanprover/lean4/issues/5002 to possibly fix this confusing behavior

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Comment by bryangingechen** [comment]
bors r+
Just in case,
bors d+

**Pull request closed without merging**
